ext/odbc/php_odbc.c

```diff
@@ -49,18 +51,140 @@
+#define CHECK_ODBC_RESULT(result) \
+	if (result->conn_ptr == NULL) { \
+		zend_throw_error(NULL, "ODBC result has already been closed"); \
+		RETURN_THROWS(); \
+	}
 
 void odbc_do_connect(INTERNAL_FUNCTION_PARAMETERS, int persistent);
+static void safe_odbc_disconnect(void *handle);
+
+static int le_pconn;
+
+static zend_class_entry *odbc_connection_ce, *odbc_result_ce;
+static zend_object_handlers odbc_connection_object_handlers, odbc_result_object_handlers;
+
+static inline odbc_link *odbc_link_from_obj(zend_object *obj) {
+	return (odbc_link *)((char *)(obj) - XtOffsetOf(odbc_link, std));
+}
+
+#define Z_ODBC_LINK_P(zv) odbc_link_from_obj(Z_OBJ_P(zv))
+#define Z_ODBC_CONNECTION_P(zv) Z_ODBC_LINK_P(zv)->connection
+
+static zend_object *odbc_connection_create_object(zend_class_entry *class_type) {
+	odbc_link *intern = zend_object_alloc(sizeof(odbc_link), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &odbc_connection_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *odbc_connection_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct ODBC\\Connection, use odbc_connect() or odbc_pconnect() instead");
+	return NULL;
+}
+
+static void odbc_link_free(odbc_link *link)
+{
+	/* If aborted via timer expiration, don't try to call any unixODBC function */
+	if (!(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+		safe_odbc_disconnect(link->connection->hdbc);
+		SQLFreeConnect(link->connection->hdbc);
+		SQLFreeEnv(link->connection->henv);
+	}
+	efree(link->connection);
+	link->connection = NULL;
+	ODBCG(num_links)--;
+	if (!link->persistent) {
+		zend_hash_del(&ODBCG(connections), link->hash);
+	}
+	zend_string_release(link->hash);
+}
+
+static void odbc_connection_free_obj(zend_object *obj)
+{
+	odbc_link *link = odbc_link_from_obj(obj);
+
+	if (!link->connection) {
+		zend_object_std_dtor(&link->std);
+		return;
+	}
+
+	odbc_link_free(link);
+	zend_object_std_dtor(&link->std);
+}
+
+static inline odbc_result *odbc_result_from_obj(zend_object *obj) {
+	return (odbc_result *)((char *)(obj) - XtOffsetOf(odbc_result, std));
+}
+
+#define Z_ODBC_RESULT_P(zv) odbc_result_from_obj(Z_OBJ_P(zv))
+
+static zend_object *odbc_result_create_object(zend_class_entry *class_type) {
+	odbc_result *intern = zend_object_alloc(sizeof(odbc_result), class_type);
 
-static int le_result, le_conn, le_pconn;
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &odbc_result_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *odbc_result_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct ODBC\\Result, use an appropriate odbc_* function instead");
+	return NULL;
+}
+
+static void odbc_result_free(odbc_result *res)
+{
+	int i;
+
+	if (res->values) {
+		for(i = 0; i < res->numcols; i++) {
```

Comment:
Could be as follows:
```suggestion
	if (res->values) {
		for (int i = 0; i < res->numcols; i++) {
```

Reply:
FYI: this is a preexisting piece of code, originally coming from `_free_odbc_result` but I'm ok with fixing this.